File: research/scicomp/Python/Control/core/pid_controller.py
"""
PID Controller Implementation
A professional PID (Proportional-Integral-Derivative) controller implementation
with advanced features including windup protection, derivative filtering,
and adaptive tuning capabilities.
"""
import numpy as np
from typing import Optional, Tuple, Union
from dataclasses import dataclass
import time
@dataclass
class PIDConfig:
    """Configuration parameters for PID controller."""
    kp: float = 1.0  # Proportional gain
    ki: float = 0.0  # Integral gain
    kd: float = 0.0  # Derivative gain
    # Anti-windup protection
    output_min: float = -np.inf
    output_max: float = np.inf
    # Derivative filtering
    derivative_filter_tau: float = 0.0  # Time constant for derivative filter
    # Sampling time
    dt: float = 0.01  # Default sampling time in seconds
class PIDController:
    """
    Professional PID Controller with advanced features.
    Features:
    - Standard PID control with configurable gains
    - Anti-windup protection
    - Derivative filtering to reduce noise sensitivity
    - Setpoint weighting for improved response
    - Reset functionality
    Examples:
        >>> config = PIDConfig(kp=2.0, ki=0.5, kd=0.1, dt=0.01)
        >>> controller = PIDController(config)
        >>> output = controller.update(setpoint=10.0, measurement=8.5)
    """
    def __init__(self, config: PIDConfig):
        """
        Initialize PID controller.
        Parameters:
            config: PID configuration parameters
        """
        self.config = config
        self.reset()
    def reset(self) -> None:
        """Reset controller internal state."""
        self._previous_error = 0.0
        self._integral = 0.0
        self._previous_measurement = 0.0
        self._previous_derivative = 0.0
        self._last_time = None
    def update(self,
               setpoint: float,
               measurement: float,
               dt: Optional[float] = None) -> float:
        """
        Compute PID control output.
        Parameters:
            setpoint: Desired value
            measurement: Current measured value
            dt: Time step (optional, uses config.dt if None)
        Returns:
            Control output
        """
        if dt is None:
            dt = self.config.dt
        # Calculate error
        error = setpoint - measurement
        # Proportional term
        proportional = self.config.kp * error
        # Integral term with anti-windup
        self._integral += error * dt
        # Anti-windup: clamp integral if output would saturate
        integral_term = self.config.ki * self._integral
        provisional_output = proportional + integral_term
        if provisional_output > self.config.output_max:
            self._integral = (self.config.output_max - proportional) / self.config.ki
        elif provisional_output < self.config.output_min:
            self._integral = (self.config.output_min - proportional) / self.config.ki
        integral = self.config.ki * self._integral
        # Derivative term (on measurement to avoid derivative kick)
        derivative_raw = -(measurement - self._previous_measurement) / dt
        # Apply derivative filtering if specified
        if self.config.derivative_filter_tau > 0:
            alpha = dt / (self.config.derivative_filter_tau + dt)
            derivative = alpha * derivative_raw + (1 - alpha) * self._previous_derivative
            self._previous_derivative = derivative
        else:
            derivative = derivative_raw
        derivative_term = self.config.kd * derivative
        # Compute total output
        output = proportional + integral + derivative_term
        # Apply output limits
        output = np.clip(output, self.config.output_min, self.config.output_max)
        # Store values for next iteration
        self._previous_error = error
        self._previous_measurement = measurement
        return output
    def get_components(self,
                      setpoint: float,
                      measurement: float,
                      dt: Optional[float] = None) -> Tuple[float, float, float]:
        """
        Get individual PID components (for analysis/tuning).
        Parameters:
            setpoint: Desired value
            measurement: Current measured value
            dt: Time step (optional)
        Returns:
            Tuple of (proportional, integral, derivative) components
        """
        if dt is None:
            dt = self.config.dt
        error = setpoint - measurement
        proportional = self.config.kp * error
        integral = self.config.ki * self._integral
        derivative_raw = -(measurement - self._previous_measurement) / dt
        if self.config.derivative_filter_tau > 0:
            alpha = dt / (self.config.derivative_filter_tau + dt)
            derivative = alpha * derivative_raw + (1 - alpha) * self._previous_derivative
        else:
            derivative = derivative_raw
        derivative_term = self.config.kd * derivative
        return proportional, integral, derivative_term
    def tune_ziegler_nichols(self,
                           ku: float,
                           tu: float,
                           method: str = 'classic') -> PIDConfig:
        """
        Auto-tune PID parameters using Ziegler-Nichols method.
        Parameters:
            ku: Ultimate gain (gain at which system oscillates)
            tu: Ultimate period (period of oscillation)
            method: Tuning method ('classic', 'pessen', 'some_overshoot', 'no_overshoot')
        Returns:
            New PIDConfig with tuned parameters
        """
        if method == 'classic':
            kp = 0.6 * ku
            ki = 2.0 * kp / tu
            kd = kp * tu / 8.0
        elif method == 'pessen':
            kp = 0.7 * ku
            ki = 2.5 * kp / tu
            kd = 0.15 * kp * tu
        elif method == 'some_overshoot':
            kp = 0.33 * ku
            ki = 2.0 * kp / tu
            kd = kp * tu / 3.0
        elif method == 'no_overshoot':
            kp = 0.2 * ku
            ki = 2.0 * kp / tu
            kd = kp * tu / 3.0
        else:
            raise ValueError(f"Unknown tuning method: {method}")
        return PIDConfig(
            kp=kp, ki=ki, kd=kd,
            dt=self.config.dt,
            output_min=self.config.output_min,
            output_max=self.config.output_max,
            derivative_filter_tau=self.config.derivative_filter_tau
        )
def simulate_pid_system(controller: PIDController,
                       plant_transfer_function,
                       setpoint: Union[float, np.ndarray],
                       duration: float,
                       dt: Optional[float] = None,
                       noise_std: float = 0.0) -> dict:
    """
    Simulate closed-loop PID control system.
    Parameters:
        controller: PID controller instance
        plant_transfer_function: System transfer function (callable)
        setpoint: Reference signal (constant or time-varying)
        duration: Simulation duration in seconds
        dt: Time step (uses controller.config.dt if None)
        noise_std: Standard deviation of measurement noise
    Returns:
        Dictionary containing time, setpoint, output, control signals
    """
    if dt is None:
        dt = controller.config.dt
    t = np.arange(0, duration, dt)
    n_points = len(t)
    # Handle setpoint
    if np.isscalar(setpoint):
        sp = np.full(n_points, setpoint)
    else:
        sp = np.array(setpoint)
        if len(sp) != n_points:
            raise ValueError("Setpoint array length must match time points")
    # Initialize arrays
    output = np.zeros(n_points)
    control = np.zeros(n_points)
    measured = np.zeros(n_points)
    # Reset controller
    controller.reset()
    # Simulation loop
    for i in range(1, n_points):
        # Add measurement noise
        measured[i-1] = output[i-1] + np.random.normal(0, noise_std)
        # Compute control signal
        control[i] = controller.update(sp[i], measured[i-1], dt)
        # Apply to plant (simplified first-order system example)
        # User should replace with actual plant model
        output[i] = plant_transfer_function(control[i], output[i-1], dt)
    return {
        'time': t,
        'setpoint': sp,
        'output': output,
        'control': control,
        'measured': measured
    }

File: research/scicomp/MATLAB/Control/core/PIDController.m
classdef PIDController < handle
    % PIDController - Professional PID controller implementation
    %
    % Features:
    % - Standard PID control with configurable gains
    % - Anti-windup protection
    % - Derivative filtering to reduce noise sensitivity
    % - Setpoint weighting for improved response
    % - Reset functionality
    %
    % Example:
    %   config = struct('kp', 2.0, 'ki', 0.5, 'kd', 0.1, 'dt', 0.01);
    %   controller = PIDController(config);
    %   output = controller.update(10.0, 8.5);
    properties (Access = private)
        config              % Configuration structure
        previous_error      % Previous error value
        integral            % Integral accumulator
        previous_measurement % Previous measurement
        previous_derivative % Previous derivative (for filtering)
    end
    methods
        function obj = PIDController(config)
            % Constructor
            %
            % Parameters:
            %   config - Configuration structure with fields:
            %            kp, ki, kd, dt, output_min, output_max,
            %            derivative_filter_tau
            validateattributes(config, {'struct'}, {'nonempty'});
            % Set default values
            defaults = struct(...
                'kp', 1.0, ...
                'ki', 0.0, ...
                'kd', 0.0, ...
                'dt', 0.01, ...
                'output_min', -Inf, ...
                'output_max', Inf, ...
                'derivative_filter_tau', 0.0 ...
            );
            % Merge with user config
            fields = fieldnames(defaults);
            for i = 1:length(fields)
                if ~isfield(config, fields{i})
                    config.(fields{i}) = defaults.(fields{i});
                end
            end
            obj.config = config;
            obj.reset();
        end
        function reset(obj)
            % Reset controller internal state
            obj.previous_error = 0.0;
            obj.integral = 0.0;
            obj.previous_measurement = 0.0;
            obj.previous_derivative = 0.0;
        end
        function output = update(obj, setpoint, measurement, dt)
            % Compute PID control output
            %
            % Parameters:
            %   setpoint - Desired value
            %   measurement - Current measured value
            %   dt - Time step (optional, uses config.dt if not provided)
            %
            % Returns:
            %   output - Control output
            if nargin < 4
                dt = obj.config.dt;
            end
            validateattributes(setpoint, {'numeric'}, {'scalar', 'finite'});
            validateattributes(measurement, {'numeric'}, {'scalar', 'finite'});
            validateattributes(dt, {'numeric'}, {'scalar', 'positive'});
            % Calculate error
            error = setpoint - measurement;
            % Proportional term
            proportional = obj.config.kp * error;
            % Integral term with anti-windup
            obj.integral = obj.integral + error * dt;
            % Anti-windup: clamp integral if output would saturate
            integral_term = obj.config.ki * obj.integral;
            provisional_output = proportional + integral_term;
            if provisional_output > obj.config.output_max
                obj.integral = (obj.config.output_max - proportional) / obj.config.ki;
            elseif provisional_output < obj.config.output_min
                obj.integral = (obj.config.output_min - proportional) / obj.config.ki;
            end
            integral = obj.config.ki * obj.integral;
            % Derivative term (on measurement to avoid derivative kick)
            derivative_raw = -(measurement - obj.previous_measurement) / dt;
            % Apply derivative filtering if specified
            if obj.config.derivative_filter_tau > 0
                alpha = dt / (obj.config.derivative_filter_tau + dt);
                derivative = alpha * derivative_raw + (1 - alpha) * obj.previous_derivative;
                obj.previous_derivative = derivative;
            else
                derivative = derivative_raw;
            end
            derivative_term = obj.config.kd * derivative;
            % Compute total output
            output = proportional + integral + derivative_term;
            % Apply output limits
            output = max(obj.config.output_min, min(obj.config.output_max, output));
            % Store values for next iteration
            obj.previous_error = error;
            obj.previous_measurement = measurement;
        end
        function [proportional, integral, derivative_term] = getComponents(obj, setpoint, measurement, dt)
            % Get individual PID components (for analysis/tuning)
            %
            % Parameters:
            %   setpoint - Desired value
            %   measurement - Current measured value
            %   dt - Time step (optional)
            %
            % Returns:
            %   proportional - Proportional component
            %   integral - Integral component
            %   derivative_term - Derivative component
            if nargin < 4
                dt = obj.config.dt;
            end
            error = setpoint - measurement;
            proportional = obj.config.kp * error;
            integral = obj.config.ki * obj.integral;
            derivative_raw = -(measurement - obj.previous_measurement) / dt;
            if obj.config.derivative_filter_tau > 0
                alpha = dt / (obj.config.derivative_filter_tau + dt);
                derivative = alpha * derivative_raw + (1 - alpha) * obj.previous_derivative;
            else
                derivative = derivative_raw;
            end
            derivative_term = obj.config.kd * derivative;
        end
        function new_config = tuneZieglerNichols(obj, ku, tu, method)
            % Auto-tune PID parameters using Ziegler-Nichols method
            %
            % Parameters:
            %   ku - Ultimate gain (gain at which system oscillates)
            %   tu - Ultimate period (period of oscillation)
            %   method - Tuning method ('classic', 'pessen', 'some_overshoot', 'no_overshoot')
            %
            % Returns:
            %   new_config - New configuration structure with tuned parameters
            if nargin < 4
                method = 'classic';
            end
            validateattributes(ku, {'numeric'}, {'scalar', 'positive'});
            validateattributes(tu, {'numeric'}, {'scalar', 'positive'});
            validatestring(method, {'classic', 'pessen', 'some_overshoot', 'no_overshoot'});
            switch method
                case 'classic'
                    kp = 0.6 * ku;
                    ki = 2.0 * kp / tu;
                    kd = kp * tu / 8.0;
                case 'pessen'
                    kp = 0.7 * ku;
                    ki = 2.5 * kp / tu;
                    kd = 0.15 * kp * tu;
                case 'some_overshoot'
                    kp = 0.33 * ku;
                    ki = 2.0 * kp / tu;
                    kd = kp * tu / 3.0;
                case 'no_overshoot'
                    kp = 0.2 * ku;
                    ki = 2.0 * kp / tu;
                    kd = kp * tu / 3.0;
            end
            new_config = obj.config;
            new_config.kp = kp;
            new_config.ki = ki;
            new_config.kd = kd;
        end
    end
end